slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -224,6 +225,109 @@ void SLAMServiceImpl::SetUpMapBuilder() {
+    }
+
+    auto image = io::Image(std::move(painted_slices.surface));
+    return image.WriteJpegToString(50);
+}
+
+bool SLAMServiceImpl::GetLatestPointCloudMapString(std::string &pointcloud) {
+    
+    cartographer::io::PaintSubmapSlicesResult painted_slices = GetLatestPaintedMapSlices();
+    auto painted_surface = painted_slices.surface.get();
+    int width = cairo_image_surface_get_width(painted_surface);
+    int height = cairo_image_surface_get_height(painted_surface);
+
+    //total number of bytes, 32 bits per pixel
+    size_t size_data = width*height*4;
+    auto data = cairo_image_surface_get_data(painted_surface);
+
+    std::vector<unsigned char> data_vect(data, data + size_data);
+    
+    std::stringbuf data_buffer;
+
+    int num_points = 0;
+
+    // Reduce resolution based off number of pixels. Output is pixels to skip
+    int scaleFactor = 4;
+    int scale = size_data/100000.f/32.f/scaleFactor;
+    
+    // Loop to filter unwanted data and reduce resolution
+    for (int i=0;i<size_data;i=i+scale*4) {
+        int rgb = 0;
+        rgb = rgb | ((int)data_vect[i+0] << 16);
+        rgb = rgb | ((int)data_vect[i+1] << 8);
+        rgb = rgb | ((int)data_vect[i+2] << 0);
+        
+        //skip pixels that are not in our map(black/past walls)
+        // this value represents [102,102,102]
+        if(rgb == 6710886)
+        continue;
+
+        //Determine probability based off color pixel
+        int prob = ViamColorToProbability((int)data_vect[i+2]);
+        
+        if(prob == 0)
+        continue;
+        
+        num_points++;
+        
+        int pixel_index = i/4;
+        int pixel_x = pixel_index%width;
+        int pixel_y = pixel_index/width;
+
+        // kPixelSize is .01, unsure of units but assumed converts to meters
+        // based on PaintSubmapSlices() and DrawPoseOnSurface()
+        float x_pos = (pixel_x-painted_slices.origin.x())*kPixelSize;
+        // Y is inverted to match output from getPosition()
+        float y_pos = -(pixel_y-painted_slices.origin.y())*kPixelSize;
+        // 2D SLAM so Z is set to 0
+        float z_pos = 0;
+
+        //rotating coordinates to match slam service expectation(XZ plane)
+        data_buffer.sputn((const char *)&x_pos, 4);
+        data_buffer.sputn((const char *)&z_pos, 4);
+        data_buffer.sputn((const char *)&y_pos, 4);
+        data_buffer.sputn((const char *)&prob, 4);
+    }
+
+    std::stringbuf pointcloud_buffer;
+    std::ostream oss(&pointcloud_buffer);
+
+    // Write our PCD file, which is written as a binary.
+    oss << "VERSION .7\n"
+        << "FIELDS x y z rgb\n"
+        << "SIZE 4 4 4 4\n"
+        << "TYPE F F F I\n"
+        << "COUNT 1 1 1 1\n"
+        << "WIDTH " << num_points << "\n"
+        << "HEIGHT " << 1 << "\n"
+        << "VIEWPOINT 0 0 0 1 0 0 0\n"
+        << "POINTS " << num_points << "\n"
+        << "DATA binary\n";
+
+    // writes data buffer to the pointcloud string
+    oss << data_buffer.str();
```

Comment:
We are technically writing the point cloud to two different buffers. The thought process for this was I would need to store the filtered pointcloud points somewhere, and I need my PCD header to know the number of points I have when I write it, so writing the header beforehand is not an option here. If there is a better or more preferred way to do this?

Reply:
I think @nicksanford solves this in his PR for orbslam. Also happy if you want to save it for the follow up PR for [RSDK-1074](https://viam.atlassian.net/browse/RSDK-1074)

Reply:
The issue is because I have to filter out pixels that are not part of my map, I would still need to copy all of my legal datapoints into something.

So I could make two vectors to hold the index and probability locations of each pixel, and then iterate those filtered pixels into a buffer, or if I should just place my map points straight into its own buffer to copy into my final buffer is something that I am not sure about.

Reply:
Took some of nicks functions and moved some things around based off nick's work.